### PR TITLE
Conditionally add a random UUID dedup ID if the queue isn't configured for content based deduplication

### PR DESCRIPTION
**Pull request by jwilmoth-nc**
## :loudspeaker: Type of change

- [X ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description

Conditionally add a random UUID dedup ID if the queue isn't configured for content based deduplication to allow consumers who are using AWS SQS's content based deduplication functionality to let AWS decide what's a duplicate.

## :bulb: Motivation and Context

Fixes an issue with the same content being delivered via a FIFO queue configured with content based deduplication.

## :green_heart: How did you test it?
Added unit test.

## :pencil: Checklist

- [ ] I reviewed submitted code
- [X ] I added tests to verify changes
- [ ] I updated reference documentation to reflect the change
- [ X] All tests passing
- [ ] No breaking changes

### CODE REVIEW

**Comment by tomazfernandes** [comment]
Thanks for the PR @jwilmoth-nc!

There seems to be a lot of formatting changes, which makes the review harder - can you please run mvn verify so the spotless plugin can apply out default format?

File: spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplate.java [hunk 1]
```diff
@@ -244,23 +250,43 @@ private Map<String, Object> addAdditionalReceiveHeaders(SqsReceiveOptionsImpl op
...
+	private <T> org.springframework.messaging.Message<T> addMissingFifoSendHeaders(String endpointName,
 			org.springframework.messaging.Message<T> message) {
-		return MessageHeaderUtils.addHeadersIfAbsent(message,
-				Map.of(SqsHeaders.MessageSystemAttributes.SQS_MESSAGE_GROUP_ID_HEADER, UUID.randomUUID().toString(),
-						SqsHeaders.MessageSystemAttributes.SQS_MESSAGE_DEDUPLICATION_ID_HEADER,
-						UUID.randomUUID().toString()));
+
+		String contentBasedDedupQueueAttribute;
+		try {
+			contentBasedDedupQueueAttribute = getQueueAttributes(endpointName).get()
```
Comment (tomazfernandes) [inline spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplate.java:269]: I tend to favor `join` over `get` - it enables cleaner exception handling. But I'm ok with `get` as well.

Also, there's a policy not to use static imports please.

File: spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplate.java [hunk 2]
```diff
@@ -408,10 +434,14 @@ private boolean isSkipAttribute(MessageSystemAttributeName name) {
...
 	private CompletableFuture<QueueAttributes> getQueueAttributes(String endpointName) {
-		return this.queueAttributesCache.computeIfAbsent(endpointName,
-				newName -> QueueAttributesResolver.builder().sqsAsyncClient(this.sqsAsyncClient).queueName(newName)
-						.queueNotFoundStrategy(this.queueNotFoundStrategy).queueAttributeNames(this.queueAttributeNames)
-						.build().resolveQueueAttributes());
+		return this.queueAttributesCache.computeIfAbsent(endpointName, newName -> {
+			// ensure we have the content based dedupe config to determine default fifo send headers
+			Set<QueueAttributeName> namesToRequest = new HashSet<>(queueAttributeNames);
+			namesToRequest.add(QueueAttributeName.CONTENT_BASED_DEDUPLICATION);
+			return QueueAttributesResolver.builder().sqsAsyncClient(this.sqsAsyncClient).queueName(newName)
```
Comment (tomazfernandes) [inline spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplate.java:441]: I think we should do this only for Fifo queues.

In order to keep responsibilities segregated, I'm thinking we can create an overload for `getQueueAttributes(String endpointName, Collection<String> additionalAttributes) ` and only call it from the Fifo methods.

How does this sound?

**Review by tomazfernandes — CHANGES_REQUESTED** [review]
@jwilmoth-nc the PR looks great, thanks!

I've left a couple of comments - please let me know if they make sense to you.

**Comment by tomazfernandes** [comment]
@maciejwalkowiak - I've asked you before but I don't quite remember what's the answer.

Whenever we run maven we get updates for S3 classes such as `AbstractCrossRegionS3Client.java`.

What should we do about those?

Seems a bit odd to update such classes in a PR related to SQS.

**Comment by maciejwalkowiak** [comment]
@tomazfernandes run `make format` afterwards. Maybe it can be automated but this is currently the only solution

File: spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplate.java (hunk 1 above)
Comment (jwilmoth-nc) [reply spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplate.java:269]: Moved to join and removed static imports from the SqsTemplate.  There were a number of pre-existing static imports that I left as-is in the test.

File: spring-cloud-aws-sqs/src/test/java/io/awspring/cloud/sqs/operations/SqsTemplateTests.java [hunk 3]
```diff
@@ -68,15 +71,19 @@
  * @author Tomaz Fernandes
  */
 @SuppressWarnings("unchecked")
+@ExtendWith(MockitoExtension.class)
```
[the remark this answers is not in the document]
Comment (jwilmoth-nc) [reply spring-cloud-aws-sqs/src/test/java/io/awspring/cloud/sqs/operations/SqsTemplateTests.java:74]: Removed.

**Comment by tomazfernandes** [comment]
@jwilmoth-nc we’ll shortly review and merge your PR, thanks.

I have one question just in case - have you confirmed targeting AWS that adding a messageDeduplicationId overrides content-based deduplication on AWS and that it does mess up deduplication?

AWS has a tendency to ignore unexpected things, so I wouldn’t be surprised if that was the case on this one.

I’m asking this because I couldn’t find anything that specific on the docs, just to measure impact.

Thanks again.

**Comment by jwilmoth-nc** [comment]
> @jwilmoth-nc we’ll shortly review and merge your PR, thanks.
> 
> I have one question just in case - have you confirmed targeting AWS that adding a messageDeduplicationId overrides content-based deduplication on AWS and that it does mess up deduplication?
> 
> AWS has a tendency to ignore unexpected things, so I wouldn’t be surprised if that was the case on this one.
> 
> I’m asking this because I couldn’t find anything that specific on the docs, just to measure impact.
> 
> Thanks again.



Hey @tomazfernandes thanks for the quick review cycle.  Yes, here's a couple of screenshots of a fifo queue with content based deduplication set and two identical messages sent w/ different MessageDeduplicationId.  Here's the AWS docs that say "[If the queue has ContentBasedDeduplication set, your MessageDeduplicationId overrides the generated one](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SendMessage.html#SQS-SendMessage-request-MessageDeduplicationId)" 

image

image

The above messages were sent with:

`aws --profile direct sqs send-message --queue-url https://sqs.us-east-1.amazonaws.com/****/spring-cloud-aws-issue-796-no-dedup.fifo --message-body "gh-issue-796" --message-group-id "test" --message-deduplication-id "1"`
and
`aws --profile direct sqs send-message --queue-url https://sqs.us-east-1.amazonaws.com/****/spring-cloud-aws-issue-796-no-dedup.fifo --message-body "gh-issue-796" --message-group-id "test" --message-deduplication-id "2"`

**Comment by tomazfernandes** [comment]
Cool, thanks for the info @jwilmoth-nc.

@maciejwalkowiak feel free to review / merge / polish anytime.

I should be able to take a closer look into this later today if that’s better.

Thanks!

**Review by tomazfernandes — APPROVED** [review]
LGTM, thanks @jwilmoth-nc!

**Pull request merged**